File: [A]tree/[A]tree-dfs-preorder/226-invert-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root
        stack = [root]
        while stack:
            node = stack.pop()
            node.left, node.right = node.right, node.left
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return root
            
# time O(n)
# space O(n)
# using tree and dfs (preorder and iterative)

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        
        def helper(node):
            if not node:
                return node
            node.left, node.right = node.right, node.left
            helper(node.left)
            helper(node.right)
            return node
        
        return helper(root)

# time O(n)
# space O(n)
# using tree and dfs (preorder and recursive)

File: [A]tree/[A]tree-dfs-preorder/226-invert-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public TreeNode invertTree(TreeNode root) {
      return helper(root);
  }

  public TreeNode helper(TreeNode node) {
      if (node == null) {
          return node;
      }
      TreeNode newLeft = node.right;
      TreeNode newRight = node.left;
      node.left = newLeft;
      node.right = newRight;
      helper(node.left);
      helper(node.right);
      return node;
  }
}

// time O(n)
// space O(n)
// using tree and dfs (preorder and recursive)